File: su-lim/W2/괄호변환.py
# myReverse : ( -> ), ) -> ( 문자 치환
def myReverse(s):
    dic = {'(':')', ')':'('}
    tmp = ""
    
    for i in s:
        tmp += dic[i]
    return tmp

# right() : 올바른 괄호 문자열 인지 판단
def right(s):
    stack = []

    for i in s:
        if i == '(':
            stack.append('(')
        else:
            try:
                stack.pop()
            except:
                return False

    if stack: # 안 비어있으면
        return False
    else:
        return True

def recursion(w):
    if not w: # empty string
        return w
    
    u = ""
    v = ""
    idx = 0
    
    # Step 2
    for idx in range(0, len(w), 2):
        u += w[idx:idx+2]
        if u.count("(") == u.count(")"):
            break
    v = w[idx+2:]
    
    # Step 3
    if right(u):
        u += recursion(v)
        return u
    
    # Step 4
    else:
        s = "("
        s += recursion(v)
        s += ")"
        s += myReverse(u[1:len(u)-1])
        return s

def solution(p):
    answer = recursion(p)
    return answer

File: hojin/week2/괄호변환.cpp
#include <string>
#include <vector>
#include <iostream>

using namespace std;

bool check_func(string u){

    vector<char> stack;

    for(auto it : u){
        if(!stack.empty() && stack.back() == '(' && it == ')'){
            stack.pop_back();
        }
        else{
            stack.push_back(it);
        }
    }
    if(!stack.empty()){
        return false;
    }
    return true;
}

string solution(string p) {

    int left = 0, right = 0, range = 0;
    bool check = false;
    string u = "", v = "", tmp = "";

    if(p == ""){
        return "";
    }

    for(auto it : p){
        if(check){
            v += it;
        }
        else{
            u += it;
            if(it == ')'){ //
                right += 1;
            }
            else{
                left += 1;
            }
            if(right == left){
                check = true;
            }
        }
    }

    if(check_func(u)){
        u += solution(v); 
    }
    else{
        tmp = "(" + solution(v) + ")";
 
        range = (int)u.size() - 1;
        for(int i=1;i<range;i++){
            if(u[i] == ')'){
                tmp += '(';
            }
            else{
                tmp += ')';
            }
        }
        return tmp;
    }

    return u;
}

int main(){

    cout << solution("()") << "\n";


    return 0;
}